apps/emqx_opentelemetry/src/emqx_otel_trace.erl

```diff
@@ -0,0 +1,275 @@
+%%--------------------------------------------------------------------
+%% Copyright (c) 2023 EMQ Technologies Co., Ltd. All Rights Reserved.
+%%
+%% Licensed under the Apache License, Version 2.0 (the "License");
+%% you may not use this file except in compliance with the License.
+%% You may obtain a copy of the License at
+%%
+%%     http://www.apache.org/licenses/LICENSE-2.0
+%%
+%% Unless required by applicable law or agreed to in writing, software
+%% distributed under the License is distributed on an "AS IS" BASIS,
+%% WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+%% See the License for the specific language governing permissions and
+%% limitations under the License.
+%%--------------------------------------------------------------------
+-module(emqx_otel_trace).
+
+-behaviour(emqx_external_trace).
+
+-export([
+    ensure_traces/1,
+    start/1,
+    stop/0
+]).
+
+-export([toggle_registered/1]).
+
+-export([
+    trace_process_publish/3,
+    start_trace_send/2,
+    end_trace_send/1,
+    event/2
+]).
+
+-include_lib("emqx/include/emqx.hrl").
+-include_lib("emqx/include/emqx_mqtt.hrl").
+-include_lib("opentelemetry_api/include/otel_tracer.hrl").
+
+-define(EMQX_OTEL_CTX, otel_ctx).
+-define(IS_ENABLED, emqx_enable).
+-define(USER_PROPERTY, 'User-Property').
+
+-define(TRACE_ALL_KEY, {?MODULE, trace_all}).
+-define(TRACE_ALL, persistent_term:get(?TRACE_ALL_KEY, false)).
+
+%%--------------------------------------------------------------------
+%% config
+%%--------------------------------------------------------------------
+
+-spec toggle_registered(boolean()) -> ok | {error, term()}.
+toggle_registered(true = _Enable) ->
+    emqx_external_trace:register_provider(?MODULE);
+toggle_registered(false = _Enable) ->
+    _ = emqx_external_trace:unregister_provider(?MODULE),
+    ok.
+
+-spec ensure_traces(map()) -> ok | {error, term()}.
+ensure_traces(#{enable := true} = Conf) ->
+    start(Conf);
```

Comment:
This seems to be called from the [`application:start/2`](https://github.com/emqx/emqx/blob/3eda1f16f1f758c883b1ec0913d056a490cbe9f2/apps/emqx_opentelemetry/src/emqx_otel_app.erl#L27) callback, but inside `start/1` here we may do [some calls](https://github.com/emqx/emqx/blob/3eda1f16f1f758c883b1ec0913d056a490cbe9f2/apps/emqx_opentelemetry/src/emqx_otel_trace.erl#L80-L81) to `application`, which may deadlock the application startup.

Reply:
This seemed safe overall, as application controller spawns another process to start an app :thinking:
 https://github.com/erlang/otp/blob/OTP-25.3.2.2/lib/kernel/src/application_controller.erl#L688
https://github.com/erlang/otp/blob/OTP-25.3.2.2/lib/kernel/src/application_controller.erl#L1385
Perhaps need to re-think it...

Reply:
As we talked about, maybe it's safe in this use case, indeed.  I'm curious about others' opinions.  :smile_cat: 
I can't recall the conditions in the past in which I bumped into this kind of deadlock, might be during the `stop` callback, which is sync.  :thinking:

Reply:
@thalesmg, following our discussion, I've updated otel lib and got rid of using `application:ensure_all_started/1` completely: https://github.com/emqx/emqx/pull/11984/commits/16407d6beb9a49aea6df185746eba321e61cc9c8
This must be a better and safer design anyway. 
Thanks for raising this concern! :raised_hands: